error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/BigDecimalRules.java

```diff
@@ -66,4 +67,92 @@ BigDecimal after(double value) {
       return BigDecimal.valueOf(value);
     }
   }
+
+  /** Prefer using {@link BigDecimal#signum()} over comparing to {@link BigDecimal#ZERO}. */
+  static final class BigDecimalCompareToZero {
+    @BeforeTemplate
+    boolean before(BigDecimal value) {
+      return Refaster.anyOf(
+          value.compareTo(BigDecimal.ZERO) == 0, BigDecimal.ZERO.compareTo(value) == 0);
+    }
+
+    @AfterTemplate
+    @AlsoNegation
+    boolean after(BigDecimal value) {
+      return value.signum() == 0;
+    }
+  }
+
+  /** Prefer using {@link BigDecimal#signum()} over more contrived alternatives. */
+  static final class BigDecimalCompareToZeroPositive {
+    @BeforeTemplate
+    boolean before(BigDecimal value) {
+      return Refaster.anyOf(
+          value.compareTo(BigDecimal.ZERO) > 0,
+          BigDecimal.ZERO.compareTo(value) < 0,
+          value.signum() > 0,
+          value.signum() >= 1,
+          value.signum() != -1);
+    }
+
+    @AfterTemplate
+    boolean after(BigDecimal value) {
+      return value.signum() == 1;
+    }
+  }
+
+  /** Prefer using {@link BigDecimal#signum()} over comparing to {@link BigDecimal#ZERO}. */
+  static final class BigDecimalCompareToZeroPositiveInclusive {
+    @BeforeTemplate
+    boolean before(BigDecimal value) {
+      return Refaster.anyOf(
+          value.compareTo(BigDecimal.ZERO) >= 0, BigDecimal.ZERO.compareTo(value) <= 0);
+    }
+
+    @AfterTemplate
+    boolean after(BigDecimal value) {
+      return value.signum() >= 0;
+    }
+  }
+
+  /** Prefer using {@link BigDecimal#signum()} over comparing to {@link BigDecimal#ZERO}. */
+  static final class BigDecimalCompareToZeroNegative {
+    @BeforeTemplate
+    boolean before(BigDecimal value) {
+      return Refaster.anyOf(
+          value.compareTo(BigDecimal.ZERO) < 0, BigDecimal.ZERO.compareTo(value) > 0);
+    }
+
+    @AfterTemplate
+    boolean after(BigDecimal value) {
+      return value.signum() == -1;
```

Comment:
(^ With this change we can also further collapse the rules using `@AfterTemplate`.)

Reply:
> Instead of == 1 and == -1, shall we settle on > 0 and < 0 respectively? 

IMO `== 1` and `== -1` are more readable as meaning 'positive sign' and 'negative sign'. The `signum` function is also [mathematically defined](https://en.wikipedia.org/wiki/Sign_function) like that (unlike comparators):
![image](https://github.com/PicnicSupermarket/error-prone-support/assets/3490872/2239f68f-6cb2-40e1-aca2-7d103f26a9f5)

Reply:
That's fair, but Java's type system still maps those numbers onto larger domain, with the alternative making clear how such won't-happen output would be treated. I.e., a trade-off between mathematical fact and programming pragmatics :shrug:

Reply:
I'm not sure if the type should be leading here; such reasoning could be extrapolated to mean you should be prepared for e.g. `Collection#size` returning something negative 😛 Perhaps this discussion signals that either `< 0` or `== -1` should be acceptable, but that other variants are not (and ofc. we still need to pick a default of how to rewrite those here).

Reply:
> such reasoning could be extrapolated to mean you should be prepared for e.g. `Collection#size` returning something negative

Luckily we can [side step](https://github.com/PicnicSupermarket/error-prone-support/blob/da5eea832975cb9222cba51935f7a22b4ff35e46/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CollectionRules.java#L32-L57) that particular case ;)

> Perhaps this discussion signals that either `< 0` or `== -1` should be acceptable, but that other variants are not (and ofc. we need to pick a default of how to rewrite those here).

Indeed, we could also drop some of the cases from the `@BeforeTemplate` body. Downside of _that_ is that we won't "clean up" mixed-use code-bases. (And it would still require agreement on what the `@AfterTemplate` should contain :sweat_smile:)

Perhaps this is a case where a flag could help, though right now Refaster doesn't support that (but the idea was also floated in #654 and #655), while a `BugChecker` approach would be _way_ more work.

Maybe I'm too a-mathematical on this point; will think about it. Also interested in @rickie's thoughts on this.

NB: Whatever we settle on: we can do the same for `{Integer,Long}#signum`. :smile:

Reply:
Upon further reflection, the signum is often multiplied; an operation that crucially depends on the [-1, 0, 1] domain. That's an argument in favour of just embracing that fact. If we do, however, it does make sense to me that we _always_ compare to these values, so also in favour of the `>= 0` and `<= 0` alternatives proposed earlier.

I pushed a PR in that direction. WDYT?

Reply:
It was my original proposal, so SGTM 😄

Reply:
> NB: Whatever we settle on: we can do the same for `{Integer,Long}#signum`. 😄

I filed #822 for this :)

Reply:
I like the argument @Venorcis made [here](https://github.com/PicnicSupermarket/error-prone-support/pull/812#discussion_r1349508629). But then you also came up with the multiplication argument that is strong.

Nice discussion 🚀 !